build-system/tasks/mocha-ci-json-reporter.js

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const fs = require('fs').promises;
+const {
+  EVENT_TEST_PASS,
+  EVENT_TEST_FAIL,
+  EVENT_RUN_END,
+  EVENT_TEST_PENDING,
+  EVENT_SUITE_BEGIN,
+  EVENT_SUITE_END,
+} = require('mocha').Runner;
+const {Base} = require('mocha').reporters;
+const {inherits} = require('mocha').utils;
+
+async function writeOutput(output, filename) {
+  try {
+    await fs.writeFile(filename, JSON.stringify(output, null, 4));
+  } catch (error) {
+    process.stdout.write(
+      Base.color('fail', `Could not write to file '${filename}'`)
+    );
+  }
+}
+
+/**
+ * Custom Mocha reporter for CI builds.
+ * Mimics the structured karma reporter, but for Karma.
+ * @param {*} runner
+ */
+function ciJsonReporter(runner) {
+  Base.call(this, runner);
+  const testResults = [];
+  let suiteList = [];
+
+  runner.on(EVENT_SUITE_BEGIN, function (suite) {
+    suiteList.push(suite.title);
+  });
+
+  runner.on(EVENT_SUITE_END, function () {
+    // We need a fresh copy every time we make a new suite,
+    // so we can't use pop here (or the suiteList info of previous
+    // tests would be destroyed)
+    suiteList = suiteList.slice(0, -1);
+  });
+
+  runner.on(EVENT_TEST_PASS, function (test) {
+    testResults.push(
+      formatForApi(test, suiteList, {
+        success: true,
+        skipped: false,
+      })
+    );
+  });
+
+  runner.on(EVENT_TEST_FAIL, function (test) {
+    testResults.push(
+      formatForApi(test, suiteList, {
+        success: false,
+        skipped: false,
+      })
+    );
+  });
+
+  runner.on(EVENT_TEST_PENDING, function (test) {
+    testResults.push(
+      formatForApi(test, suiteList, {
+        success: false,
+        skipped: true,
+      })
+    );
+  });
+
+  runner.on(EVENT_RUN_END, async function () {
+    // Apparently we'll need to add a --no-exit flag when calling this
+    // to allow for the asynchronous reporter.
+    // See https://github.com/mochajs/mocha/issues/812
```

Comment:
Can we assert this flag is provided and complain if it is not, providing some explanation?

Reply:
I don't know if we can -- I don't know if using that flag is a good idea or not, I think I'd like a few more eyes on this.